### PR TITLE
Creates DiscreteMaterial object to enable performing iterative propert…

### LINKED ISSUE #5166: Add ability to recompute Materials at a quadrature point
`Material` will add the following methods:
- `computeProperty(unsigned int qp)` - this will set _qp to qp and call `computeQpProperties()
- `unsigned int id materialProperty(std::string prop_name)` - this will take a material property name and return a unique ID for the property name
- `recomputeMaterial(unsigned int id, unsigned int qp)` recomputes the necessary `Material` objects
at the provided quadrature point

### PR DESCRIPTION
**Pull request by aeslaughter**
This is nearly the same as my previous attempt to have materials call compute on other materials, but the calling of the recompute method is explicitly performed by retrieving a special type of Material object, a `DiscreteMaterial`, which is in keeping with the recently added `DiscreteElementUserObject`. This design was suggested by @permcody and @sapitts.

(closes #5166)

### CODE REVIEW

File: framework/include/base/DependencyResolverInterface.h [hunk 1]
```diff
@@ -72,6 +90,9 @@ DependencyResolverInterface::sort(typename std::vector<T> & vector)
 
   for (typename std::vector<T>::iterator iter = start; iter != end ; ++iter)
   {
+    if ((*iter)->allowCyclicDepenedency())
+      continue;
```
Comment (permcody) [inline framework/include/base/DependencyResolverInterface.h]: I don't think I like any of these changes regarding the DependencyResolveInterface. 

Consider all of the changes this caused.
- You're adding `validParams` to this interface where there was none before. 
- Changing the constructor to accept a Parameter object.
- Changing several classes that inherit from this interface to provide Parameters
- Augmenting all of those interfaces with a parameter that shouldn't be set by most of them
- Using that parameter to bypass the entire purpose of this class
- Still sorting based on potentially incomplete or bad information (what is the behavior of sorting with cyclic dependancies)

This is simply the wrong design here.

Perhaps the DiscreteMaterial shouldn't participate or use the DependencyResolver Interface, or if you DO need it (you might), maybe it shouldn't use the same Resolver because it is allowed to have cyclic dependencies. Basically if it isn't sorting correctly, why is it in here at all?


You're adding Parameters to the interface so you can access a specific set a Boolean to basically turn off the whole

File: framework/include/base/MooseObjectWarehouseBase.h [hunk 2]
```diff
@@ -242,6 +244,25 @@ MooseObjectWarehouseBase<T>::addObject(MooseSharedPointer<T> object, THREAD_ID t
...
+  typename std::vector<MooseSharedPointer<T> >::const_iterator it;
+  for (it = _all_objects[tid].begin(); it != _all_objects[tid].end(); ++it)
+  {
+    if ((*it)->name() == name)
+    {
+      output = *it;
+      break;
+    }
+  }
+  return output;
```
Comment (permcody) [inline framework/include/base/MooseObjectWarehouseBase.h]: Looks like it's possible to return an empty pointer. If that's the design you should probably handle that in the callers below, else you should throw an error here.

File: framework/include/materials/Material.h [hunk 3]
```diff
@@ -155,6 +156,16 @@ class Material :
...
   getSuppliedItems() { return _supplied_props; }
 
+
+  /**
+   * Return a set of propreties that have been accessed with materialProperty
+   * @return A reference to the properties with calls to materialProperty
+   */
+   //virtual
+   //const std::set<std::string> &
+   //getIterativeItems() { return _iterative_props; }
```
Comment (permcody) [inline framework/include/materials/Material.h]: Delete this

File: framework/src/base/DependencyResolverInterface.C [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+InputParameters validParams<DependencyResolverInterface>()
+{
+  InputParameters params = emptyInputParameters();
+  params.addParam<bool>("allow_cyclic_dependency", false, "Allows for cyclic dependencies to exists for this object (object is also not sorted).");
+  return params;
+}
+
+
+DependencyResolverInterface::DependencyResolverInterface(const InputParameters & parameters) :
+    _allow_cyclic(parameters.get<bool>("allow_cyclic_dependency"))
```
Comment (permcody) [inline framework/src/base/DependencyResolverInterface.C]: Basically this means, don't actually resolve dependencies.

File: framework/src/base/FEProblem.C [hunk 5]
```diff
@@ -1575,6 +1577,34 @@ FEProblem::projectSolution()
...
 
 
+MooseSharedPointer<DiscreteMaterial>
+FEProblem::getDiscreteMaterial(std::string name, Moose::MaterialDataType type, THREAD_ID tid)
+{
+  MooseSharedPointer<Material> output;
+  switch (type)
+  {
+  case Moose::BLOCK_MATERIAL_DATA:
+    output = _materials.getObject(name, tid);
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: Could be empty here - see comment above.

File: framework/src/base/FEProblem.C [hunk 6]
```diff
@@ -1575,6 +1577,34 @@ FEProblem::projectSolution()
...
+  case Moose::BOUNDARY_MATERIAL_DATA:
+  case Moose::FACE_MATERIAL_DATA:
+    name += "_face";
+    output = _face_materials.getObject(name, tid);
+    break;
+  }
+
+  MooseSharedPointer<DiscreteMaterial> discrete = MooseSharedNamespace::dynamic_pointer_cast<DiscreteMaterial>(output);
+  if (!discrete)
+    mooseError("Unable to locate the a DiscreteMaterial with the name " << name);
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: Ah - so here's where we'll hit the mooseError(). You might want to document that it's possible to get an empty pointer though in your method.

File: framework/src/base/FEProblem.C [hunk 7]
```diff
@@ -3917,7 +3947,7 @@ FEProblem::checkDependMaterialsHelper(const std::map<SubdomainID, std::vector<Mo
       std::ostringstream oss;
       oss << "One or more Material Properties were not supplied on block " << j->first << ":\n";
       for (std::set<std::string>::iterator i = difference.begin(); i != difference.end();  ++i)
-        oss << *i << "\n";
+      oss << *i << "\n";
```
Comment (permcody) [inline framework/src/base/FEProblem.C]: wrong indentation

File: framework/src/materials/DiscreteMaterial.C [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+/****************************************************************/
+
+// MOOSE includes
+#include "DiscreteMaterial.h"
+
+template<>
+InputParameters validParams<DiscreteMaterial>()
+{
+  InputParameters params = validParams<Material>();
+  params.set<bool>("allow_cyclic_dependency") = true;
```
Comment (permcody) [inline framework/src/materials/DiscreteMaterial.C]: I think you should have a Material BASE CLASS. You'll inherit `Material` and `DiscreteMaterial from that`. Then you won't need to mess with the Resolver at all for `DiscreteMaterial`.

File: test/tests/materials/recompute/tests [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    type = 'Exodiff'
+    input = 'recompute_no_calc.i'
+    exodiff = 'recompute_no_calc_out.e'
+  [../]
+
+  [./newton_error]
+    # Test error message when trying to get a non DiscreteMaterial object
+    type = 'RunException'
+    input = 'recompute_error.i'
+    expect_err = 'nable to locate the a DiscreteMaterial with the name recompute_props'
```
Comment (permcody) [inline test/tests/materials/recompute/tests]: Just curious about this expect_err. Are you skipping the first character because you don't want to worry about case? These are Regex's so you can just do `[Uu]nable to locate...`

File: test/src/userobjects/RestartableTypes.C [hunk 10]
```diff
@@ -17,7 +17,7 @@
 template<>
 InputParameters validParams<RestartableTypes>()
 {
-  InputParameters params = validParams<UserObject>();
+  InputParameters params = validParams<GeneralUserObject>();
```
Comment (permcody) [inline test/src/userobjects/RestartableTypes.C]: These fixes don't appear to be related to this PR. Is this left over stuff from your Warehouse refactor?

File: framework/include/base/MooseObjectWarehouseBase.h (hunk 2 above)
Comment (aeslaughter) [inline framework/include/base/MooseObjectWarehouseBase.h]: This method should be getActiveObject to go with the hasActiveObject method, I will fix it.

File: framework/src/base/FEProblem.C (hunk 5 above)
Comment (aeslaughter) [inline framework/src/base/FEProblem.C]: This should fail if empty, I will add that.

File: framework/include/base/DependencyResolverInterface.h (hunk 1 above)
Comment (aeslaughter) [inline framework/include/base/DependencyResolverInterface.h]: The first three bullets actually makes this interface more consistent with other interfaces, most have parameters.

Bypassing the dependency check was the simpliest way to get the `DiscreteMaterials` to work as intended without doing one of the following:
    - creating a separate storage structure for `DiscreteMaterial` objects, but doing this will require creating separate `MaterialData` objects so that all the material property checks can be handled properly. 
   - another option is to create separate storage contains, but re-join them for the depedency checks, but this involves alot of `if (discrete)` stuff all over the place. I started down this road initially and it just seemed messy. At this point the `DiscreteMaterial` should just be its own system, but that leads back to the first issue of MaterialData.

Comment (aeslaughter) [inline framework/include/base/DependencyResolverInterface.h]: There is another option, in my previous attempt there was a bunch of new logic added to the dependency interface for "iterative properties", I could restore this stuff and have DiscreteMataial use the iterative stuff.

File: test/tests/materials/recompute/tests (hunk 9 above)
Comment (aeslaughter) [inline test/tests/materials/recompute/tests]: Just copy/paste error.

File: test/src/userobjects/RestartableTypes.C (hunk 10 above)
Comment (aeslaughter) [inline test/src/userobjects/RestartableTypes.C]: They are related. By adding parameters to the interface, which is on GeneralUserObject, these objects miss getting the new parameter.

Even if we don't add the parameter to the interface in the end, these should be fixed.
